fuel-merkle/src/sparse/proof.rs

```diff
@@ -0,0 +1,795 @@
+        }
+    }
+
+    pub fn is_inclusion(&self) -> bool {
+        match self {
+            Proof::Inclusion(_) => true,
+            Proof::Exclusion(_) => false,
+        }
+    }
+
+    pub fn is_exclusion(&self) -> bool {
+        !self.is_inclusion()
+    }
+}
+
+#[derive(Clone, Eq, PartialEq)]
+pub struct InclusionProof {
+    pub proof_set: ProofSet,
+}
+
+impl InclusionProof {
+    pub fn verify(&self, root: &Bytes32, key: &MerkleTreeKey, value: &[u8]) -> bool {
+        let Self { proof_set } = self;
+        let leaf = Node::create_leaf(key.as_ref(), value);
+        let mut current = *leaf.hash();
+        for (i, side_hash) in proof_set.iter().enumerate() {
+            let index = u32::try_from(proof_set.len() - 1 - i).expect("Index is valid");
+            let prefix = Prefix::Node;
+            current = match key.get_instruction(index).expect("Infallible") {
+                Instruction::Left => Node::calculate_hash(&prefix, &current, side_hash),
+                Instruction::Right => Node::calculate_hash(&prefix, side_hash, &current),
+            };
+        }
+        current == *root
+    }
+}
+
+impl Debug for InclusionProof {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let proof_set = self.proof_set.iter().map(hex::encode).collect::<Vec<_>>();
+        f.debug_struct("InclusionProof")
+            .field("Proof set", &proof_set)
+            .finish()
+    }
+}
+
+#[derive(Clone, Eq, PartialEq)]
+pub struct ExclusionLeaf {
+    leaf_key: Bytes32,
+    leaf_value: Bytes32,
+}
+
+impl ExclusionLeaf {
+    fn leaf_key(&self) -> &Bytes32 {
+        &self.leaf_key
+    }
+
+    fn leaf_value(&self) -> &Bytes32 {
+        &self.leaf_value
+    }
+
+    fn is_placeholder(&self) -> bool {
+        self.leaf_value() == zero_sum()
+    }
+
+    fn hash(&self) -> Bytes32 {
+        if self.is_placeholder() {
+            *zero_sum()
+        } else {
+            Node::calculate_hash(&Prefix::Leaf, self.leaf_key(), self.leaf_value())
+        }
+    }
+}
+
+impl From<Node> for ExclusionLeaf {
+    fn from(node: Node) -> Self {
+        ExclusionLeaf {
+            leaf_key: *node.leaf_key(),
+            leaf_value: *node.leaf_data(),
+        }
+    }
+}
+
+impl Debug for ExclusionLeaf {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("ExclusionLeaf")
+            .field("Leaf key", &hex::encode(self.leaf_key))
+            .field("Leaf value", &hex::encode(self.leaf_value))
+            .finish()
+    }
+}
+
+#[derive(Clone, Eq, PartialEq)]
+pub struct ExclusionProof {
+    pub proof_set: ProofSet,
+    pub leaf: ExclusionLeaf,
+}
+
+impl ExclusionProof {
+    pub fn verify(&self, root: &Bytes32, key: &MerkleTreeKey) -> bool {
+        let Self { proof_set, leaf } = self;
+        if !leaf.is_placeholder() && *leaf.leaf_key() == key.as_ref() {
+            return false;
+        }
+        let mut current = leaf.hash();
```

Comment:
We want to process enum here. If the `leaf` is a `(key, value)`, then we want to use `Node::create_leaf(key, value)`. Otherwise we want to use `Node::create_placeholder()`.

After you can call `node.hash()`.